src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java

```diff
@@ -35,9 +38,41 @@ public Object filter(
     Object[] args,
     Map<String, Object> kwargs
   ) {
-    return new ArrayList<>(
-      Sets.intersection(objectToSet(var), objectToSet(parseArgs(interpreter, args)))
-    );
+    Object argObj = parseArgs(interpreter, args);
+
+    Set<Object> varSet = objectToSet(var);
+    Set<Object> argSet = objectToSet(argObj);
+
+    boolean areMismatchedElementTypes = !getTypeOfSetElements(varSet)
+      .equals(getTypeOfSetElements(argSet));
+
+    if (areMismatchedElementTypes) {
+      interpreter.addError(
+        new TemplateError(
+          TemplateError.ErrorType.WARNING,
+          TemplateError.ErrorReason.OTHER,
+          TemplateError.ErrorItem.FILTER,
+          String.format(
+            "Mismatched types. `value` elements are of type `%s` and `list` elements are of type `%s`. This may lead to unexpected behavior.",
+            getTypeOfSetElements(varSet),
+            getTypeOfSetElements(argSet)
+          ),
+          "list",
+          interpreter.getLineNumber(),
+          -1,
+          null
+        )
+      );
+    }
+
+    return new ArrayList<>(Sets.intersection(varSet, argSet));
+  }
+
+  private String getTypeOfSetElements(Set<Object> set) {
+    if (set.isEmpty()) {
+      return "null";
```

Comment:
Should we make this optional? You should be able to intersect a set with an empty set, however it appears the above logic will add a warning for that.

Reply:
Kept this existing logic, but reworked the conditional check